### PR TITLE
ci: change to use Jenkins

### PR DESCRIPTION
**Pull request by kuisathaverat**
A new Jenkins job to replace the Travis CI and appveyor.

### CODE REVIEW

**Review by mdelapenya — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env groovy
+
+@Library('apm@current') _
+
+pipeline {
+  agent { label 'ubuntu-20 && immutable' }
+  environment {
+    BASE_DIR="src/github.com/elastic/gosigar"
```
Comment (v1v) [inline .ci/Jenkinsfile:8]:
```suggestion
    REPO = 'gosigar'
    BASE_DIR = "src/github.com/elastic/${env.REPO}"
```

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    durabilityHint('PERFORMANCE_OPTIMIZED')
+    rateLimitBuilds(throttle: [count: 60, durationName: 'hour', userBoost: true])
+    quietPeriod(10)
+  }
+  triggers {
+    issueCommentTrigger('(?i)(.*(?:jenkins\\W+)?run\\W+(?:the\\W+)?tests(?:\\W+please)?|/test).*')
+  }
+  parameters {
+    booleanParam(name: 'run_all_stages', defaultValue: false, description: 'Force to run all stages.')
+  }
```
Comment (v1v) [inline .ci/Jenkinsfile:26]:
```suggestion
```

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+def getExamplesDirs(){
+  return [
+    "examples/df",
+    "examples/free",
+    "examples/ps",
+    "examples/uptime"
+  ]
+}
+
+def  buildExamples(){
```
Comment (v1v) [inline .ci/Jenkinsfile:106]:
```suggestion
def buildExamples(){
```

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  getExamplesDirs().each { exampleDir ->
+      cmd(label: "Build Example ${exampleDir}", script: "go build -o ${exampleDir}/out.exe ./${exampleDir}")
```
Comment (v1v) [inline .ci/Jenkinsfile:108]: Is the .exe extension correct for the linux builds? Or it does not matter?

File: .ci/jobs/main.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    project-type: multibranch
+    script-path: .ci/Jenkinsfile
+    scm:
+      - github:
+          branch-discovery: no-pr
+          discover-pr-forks-strategy: merge-current
+          discover-pr-forks-trust: permission
+          discover-pr-origin: merge-current
+          discover-tags: true
+          head-filter-regex: '(master|main|PR-.*)'
```
Comment (v1v) [inline .ci/jobs/main.yml:16]: This will filter any tags, if so, should the above line set to false?

File: .ci/Jenkinsfile (hunk 4 above)
Comment (kuisathaverat) [reply .ci/Jenkinsfile:108]: in order to use the same code on all platforms and it simpler, I have used the extension `.exe`,  on Linux the extension does not matter.

File: .ci/jobs/main.yml (hunk 5 above)
Comment (kuisathaverat) [inline .ci/jobs/main.yml:16]:
```suggestion
          head-filter-regex: '(master|main|PR-.*|v\d+\.\d+\.\d+)'
```

File: .ci/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        }
+      }
+    }
+    stage('Build and test'){
+      matrix {
+        agent {label "${PLATFORM}"}
+        axes {
+          axis {
+            name 'PLATFORM'
+            values 'macOS', 'ubuntu-20 && immutable', 'windows-2012-r2 && immutable'
```
Comment (kuisathaverat) [inline .ci/Jenkinsfile:52]:
```suggestion
            values 'macosx', 'ubuntu-20 && immutable', 'windows-2012-r2 && immutable'
```

**Comment by kuisathaverat** [comment]
I will say that does not build on Go 1.11, it fails on all platforms with a dependency missing error.

```
[2021-02-03T16:00:40.704Z] + go build
[2021-02-03T16:00:40.704Z] /var/lib/jenkins/workspace/Beats_gosigar_PR-153/src/github.com/elastic/gosigar/sigar_unix.go:11:2: cannot find package "golang.org/x/sys/unix" in any of:
[2021-02-03T16:00:40.704Z] 	/var/lib/jenkins/workspace/Beats_gosigar_PR-153/.gvm/versions/go1.11.darwin.amd64/src/golang.org/x/sys/unix (from $GOROOT)
[2021-02-03T16:00:40.704Z] 	/var/lib/jenkins/workspace/Beats_gosigar_PR-153/src/golang.org/x/sys/unix (from $GOPATH)
```

**Comment by fearful-symmetry** [comment]
Is there any reason we're using go.11 here?  Is it standard to test against a release that old?

**Review by v1v — APPROVED** [review]

**Comment by kuisathaverat** [comment]
> Is there any reason we're using go.11 here? Is it standard to test against a release that old?

It is in the Travis-CI pipeline https://github.com/elastic/gosigar/blob/master/.travis.yml#L8, I guess we can rid of it

**Comment by fearful-symmetry** [comment]
@kuisathaverat Unless we have some specific backwards compatibility matrix requirements here, I would lean towards removing it for now. gosigar has a limited shelf life right now, and we're trying to deprecate it in the next few releases, and beats is built against 1.15.7.

File: .ci/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        agent {label "${PLATFORM}"}
+        axes {
+          axis {
+            name 'PLATFORM'
+            values 'macosx', 'ubuntu-20 && immutable', 'windows-2012-r2 && windows-immutable'
+          }
+          axis {
+            name 'GO_VERSION'
+            values '1.11', '1.15'
+          }
```
Comment (kuisathaverat) [inline .ci/Jenkinsfile:57]:
```suggestion
```

**Comment by kuisathaverat** [comment]
@fearful-symmetry to avoid having to change the Go version on the project every time we bump it, we will use the default Go version we set for all projects https://github.com/elastic/apm-pipeline-library/blob/master/vars/goDefaultVersion.groovy#L41

**Comment by kuisathaverat** [comment]
there is a small bug on the library fixed on https://github.com/elastic/apm-pipeline-library/pull/954 , once we merge that PR and release a new Library version we can test this again and merge the PR.

**Pull request merged**